sky/clouds/gcp.py

```diff
@@ -524,6 +524,81 @@ def get_vcpus_mem_from_instance_type(
         return service_catalog.get_vcpus_mem_from_instance_type(instance_type,
                                                                 clouds='gcp')
 
+    def get_available_reservation_resources(
+        self,
+        instance_type: str,
+        region: str,
+        zone: Optional[str],
+        specific_reservations: Set[str],
+    ) -> int:
+        del region  # Unused
+        assert zone is not None, 'GCP requires zone to get available reservations.'
+        reservations = self._list_reservations_for_instance_type_in_zone(
+            instance_type, zone)
+        return self._count_available_resources_for_reservations(
+            reservations, specific_reservations)
+
+    def _list_reservations_for_instance_type_in_zone(
+        self,
+        instance_type: str,
+        zone: Optional[str],
+    ) -> List[Dict]:
+        list_reservations_cmd = (
+            'gcloud compute reservations list '
+            f'--filter="zone={zone} AND specificReservation.instanceProperties.machineType={instance_type} AND status=READY" '
+            '--format="json(specificReservation.count, specificReservation.inUseCount, specificReservationRequired, selfLink)"'
+        )
+        returncode, stdout, stderr = subprocess_utils.run_with_retries(
+            list_reservations_cmd,
+            retry_returncode=[255],
+        )
+        subprocess_utils.handle_returncode(
+            returncode,
+            list_reservations_cmd,
+            error_msg=
+            f'Failed to get list reservations for {instance_type!r} in {zone!r}.',
+            stderr=stderr,
+            stream_logs=True,
+        )
+        return json.loads(stdout)
```

Comment:
Since we are calling this method for every zone, the optimization time becomes very long when the `specific_reservation` is specified, as we are calling the gcloud CLI many times, each of which contributes several seconds to the delay. One optimization we can do could be something like the following:

```suggestion
    def get_available_reservation_resources(
        self,
        instance_type: str,
        region: str,
        zone: Optional[str],
        specific_reservations: Set[str],
    ) -> int:
        del region  # Unused
        assert zone is not None, 'GCP requires zone to get available reservations.'
        reservations = self._list_reservations_for_instance_type(
            instance_type, zone)
        filtered_reservations = []
        for reservation in reservations:
            if reservation['zone'].endswith(zone):
                filtered_reservations.append(reservation)
        return self._count_available_resources_for_reservations(
            filtered_reservations, specific_reservations)

    @functools.lru_cache()
    def _list_reservations_for_instance_type(
        self,
        instance_type: str,
    ) -> List[Dict]:
        list_reservations_cmd = (
            'gcloud compute reservations list '
            f'--filter="specificReservation.instanceProperties.machineType={instance_type} AND status=READY" '
            '--format="json(specificReservation.count, specificReservation.inUseCount, specificReservationRequired, selfLink, zone)"'
        )
        returncode, stdout, stderr = subprocess_utils.run_with_retries(
            list_reservations_cmd,
            retry_returncode=[255],
        )
        subprocess_utils.handle_returncode(
            returncode,
            list_reservations_cmd,
            error_msg=
            f'Failed to get list reservations for {instance_type!r}.',
            stderr=stderr,
            stream_logs=True,
        )
        return json.loads(stdout)
```

This reduces the optimization time from more than 160 seconds to 6 seconds for me with a single specific reservation with `n2-standard-2`. To reproduce, use the following scripts:

```yaml
# task.yaml
resources:
  cloud: gcp
  cpus: 2
  specific_reservations:
    - projects/skypilot-xxx/zones/us-central1-c/reservations/specific-reservation-test

setup: |
  echo test

run: |
  echo test
```

```bash
time python -u - <<EOF
import sky
with sky.Dag() as d:
  task = sky.Task.from_yaml('task.yaml')

sky.optimize(d)
EOF
```